File: ghidra_scripts/export_asm.py
# Ghidra post-analysis script (Jython)
# ML-ready disassembly exporter for GraphCodeBERT pipeline

import os

# Check if running in Ghidra environment
try:
    args = getScriptArgs()
except NameError:
    print("ERROR: This script must be run from within Ghidra using analyzeHeadless.bat")
    print("Usage: analyzeHeadless.bat [project] [project_name] -import [binary] -postScript ExportAsm.java [output_path]")
    import sys
    sys.exit(1)

if len(args) < 1:
    println("ERROR: No output path provided as script argument.")
else:
    output_path = args[0]

    program = currentProgram
    listing = program.getListing()
    fm = program.getFunctionManager()

    # Ensure output directory exists
    parent = os.path.dirname(output_path)
    if parent and not os.path.isdir(parent):
        os.makedirs(parent)

    # =========================
    # NORMALIZATION FUNCTIONS
    # =========================
    def normalize_register(reg):
        if reg is None:
            return "REG"

        name = reg.getName().lower()

        if name in ["rsp", "esp"]:
            return "SP"
        if name in ["rbp", "ebp"]:
            return "BP"

        size = reg.getBitLength()

        if size == 64:
            return "REG64"
        if size == 32:
            return "REG32"
        if size == 16:
            return "REG16"
        if size == 8:
            return "REG8"

        return "REG"

    def normalize_operand(obj):
        if obj is None:
            return "UNK"

        # Register
        if isinstance(obj, Register):
            return normalize_register(obj)

        # Immediate
        if isinstance(obj, Scalar):
            return "IMM"

        # Memory reference (Address objects often appear here)
        try:
            if obj.toString().startswith("0x"):
                return "MEM"
        except:
            pass

        return str(obj)

    # =========================
    # OUTPUT
    # =========================
    f = open(output_path, "w")

    try:
        f.write("; Disassembly of: %s\n" % program.getName())
        f.write("; Format: %s\n" % program.getExecutableFormat())
        f.write("; Architecture: %s\n" % program.getLanguageID())
        f.write("; Functions: %d\n\n" % fm.getFunctionCount())

        func_count = 0
        instr_count = 0

        functions = fm.getFunctions(True)

        for func in functions:

            entry = func.getEntryPoint()
            f.write("; --- Function: %s (0x%s) ---\n" % (func.getName(), entry))

            body = func.getBody()
            instr_iter = listing.getInstructions(body, True)

            for instr in instr_iter:

                addr = instr.getAddress()
                mnemonic = instr.getMnemonicString()
                num_ops = instr.getNumOperands()

                ops_list = []

                for i in range(num_ops):
                    objs = instr.getOpObjects(i)

                    if objs is None:
                        continue

                    norm = " ".join([normalize_operand(o) for o in objs])
                    ops_list.append(norm)

                ops = " | ".join(ops_list)

                f.write("0x%s:\t%s\t%s\n" % (str(addr), mnemonic, ops))
                instr_count += 1

            f.write("\n")
            func_count += 1

        println("Exported %d functions (%d instructions) -> %s"
                % (func_count, instr_count, output_path))

    finally:
        f.close()

File: ghidra_scripts/ExportAsm.java
import ghidra.app.script.GhidraScript;
import ghidra.program.model.listing.*;
import ghidra.program.model.address.*;
import ghidra.program.model.scalar.Scalar;
import ghidra.program.model.lang.Register;

import java.io.*;
import java.util.*;

/**
 * Ghidra post-analysis script (Java)
 * Exports normalized assembly code for ML processing.
 * 
 * USAGE:
 * analyzeHeadless.bat [project] [project_name] -import [binary] \
 *   -postScript "ExportAsm.java" "[output_path.asm]" \
 *   -deleteProject
 * 
 * EXAMPLE:
 * analyzeHeadless.bat C:/ghidra/projects MyProject -import malware.exe \
 *   -postScript "ExportAsm.java" "output/asm.asm" \
 *   -deleteProject
 */
public class ExportAsm extends GhidraScript {

    // =========================
    // REGISTER NORMALIZATION
    // =========================
    private String normalizeRegister(Register reg) {
        if (reg == null) return "REG";

        String name = reg.getName().toLowerCase();

        if (name.equals("rsp") || name.equals("esp")) return "SP";
        if (name.equals("rbp") || name.equals("ebp")) return "BP";

        int size = reg.getBitLength();

        if (size == 64) return "REG64";
        if (size == 32) return "REG32";
        if (size == 16) return "REG16";
        if (size == 8) return "REG8";

        return "REG";
    }

    // =========================
    // OPERAND NORMALIZATION
    // =========================
    private String normalizeOperandObjects(Object[] objs) {
        if (objs == null || objs.length == 0) return "";

        List<String> parts = new ArrayList<>();

        for (Object obj : objs) {

            if (obj instanceof Register) {
                parts.add(normalizeRegister((Register) obj));
            }

            else if (obj instanceof Scalar) {
                parts.add("IMM");
            }

            else if (obj instanceof Address) {
                parts.add("MEM");
            }

            else {
                parts.add("UNK");
            }
        }

        return String.join(" ", parts);
    }

    @Override
    public void run() throws Exception {

        String[] args = getScriptArgs();
        if (args.length < 1) {
            println("ERROR: No output path provided as script argument.");
            return;
        }

        String outputPath = args[0];

        File outFile = new File(outputPath);
        File parentDir = outFile.getParentFile();
        if (parentDir != null && !parentDir.exists()) {
            parentDir.mkdirs();
        }

        Program program = currentProgram;
        Listing listing = program.getListing();
        FunctionManager fm = program.getFunctionManager();

        PrintWriter writer = new PrintWriter(new BufferedWriter(new FileWriter(outFile)));

        try {
            writer.println("; Disassembly of: " + program.getName());
            writer.println("; Format: " + program.getExecutableFormat());
            writer.println("; Architecture: " + program.getLanguageID());
            writer.println("; Functions: " + fm.getFunctionCount());
            writer.println();

            int funcCount = 0;
            int instrCount = 0;

            FunctionIterator functions = fm.getFunctions(true);

            while (functions.hasNext()) {

                if (monitor.isCancelled()) break;

                Function func = functions.next();
                Address entry = func.getEntryPoint();

                writer.println("; --- Function: " + func.getName()
                        + " (0x" + entry + ") ---");

                AddressSetView body = func.getBody();

                InstructionIterator instrIter = listing.getInstructions(body, true);

                while (instrIter.hasNext()) {

                    Instruction instr = instrIter.next();

                    Address addr = instr.getAddress();
                    String mnemonic = instr.getMnemonicString();

                    int numOps = instr.getNumOperands();
                    StringBuilder ops = new StringBuilder();

                    for (int i = 0; i < numOps; i++) {

                        if (i > 0) ops.append(" | ");

                        Object[] opObjs = instr.getOpObjects(i);
                        ops.append(normalizeOperandObjects(opObjs));
                    }

                    writer.println("0x" + addr + ":\t"
                            + mnemonic + "\t"
                            + ops.toString());

                    instrCount++;
                }

                writer.println();
                funcCount++;
            }

            println("Exported " + funcCount + " functions ("
                    + instrCount + " instructions) -> " + outputPath);

        } finally {
            writer.close();
        }
    }
}